### ISSUE TITLE
flow_from_directory produces batches of varying size

### ISSUE THREAD

**Issue by lhk** [fchollet/keras#5406]
I'm using the builtin image preprocessing to produce a stream of batches.
In the directory, there is one subfolder (=class) with a total of 8 pictures.

If I specify a `batch_size` >= 8, every batch has the same size.
But if the `batch_size` is <8, then the batch_sizes will be:
 - `batch_size`
 - `8%batch_size`
 - `batch_size`
 - `8%batch_size`
 - ...

so for example: 6, 2, 6, 2, 6

This seems like a bug to me.
It might make sense, if Keras is internally keeping track of the images it has already visited.
But even then, this behaviour should be documented.

**Comment by lhk** [comment]
I've taken a look at the sourcecode: https://github.com/fchollet/keras/blob/master/keras/preprocessing/image.py

The `DirectoryIterator` does the following:
 1. look at a directory, scan for subdirectories (=classes)
 2. count the files in the subdirectories = number of samples
 3. create an array of filenames. length is the number of samples

Then it uses an Iterator to choose random indices and load the corresponding images.
The `flow_index` method is at line 637, here is some code:

        ```
while 1:
            if seed is not None:
                np.random.seed(seed + self.total_batches_seen)
            if self.batch_index == 0:
                index_array = np.arange(n)
                if shuffle:
                    index_array = np.random.permutation(n)

            current_index = (self.batch_index * batch_size) % n
            if n >= current_index + batch_size:
                current_batch_size = batch_size
                self.batch_index += 1
            else:
                current_batch_size = n - current_index
                self.batch_index = 0
            self.total_batches_seen += 1
            yield (index_array[current_index: current_index + current_batch_size],
                   current_index, current_batch_size)
```

As you can see, the indices are chosen at random, if the parameter `shuffle` is `True`. It is `True` by default.

```
index_array = np.arange(n)
if shuffle:
    index_array = np.random.permutation(n)
```

And the slicing in the `index_array` is done with a `current_batch_size`.
Which is smaller than the specified `batch_size` after `number_samples // batch_size` iterations, if `number_samples % batch_size != 0.`

I think this is broken.
The default behaviour should be sensible. And if the indices are chosen at random, it doesn't make sense to have a varying `batch_size`.

For the Keras training routines, this doesn't really matter.
But as soon as you plug the Keras generators in a pipeline of your own, which relies on a constant `batch_size`, this produces problems.

**Comment by unrealwill** [comment]
Hello,

It seems to be correct intended behaviour.
The random shuffling is done "without replacement" so that each image is seen only one time.
You probably want to create a sampling "with replacement" flow for your case if you need a fixed current_batch_size, or you can sample without replacement and skip the last smaller batch.

**Comment by edgimar** [comment]
Using Keras 2.0.4, I have noticed that for the "last" batch that `flow_from_directory` produces X and y whose first dimension length doesn't match each other.  Haven't looked into it further, but this seems like a bug to me.

More generally, the "correct intended behavior" doesn't seem very useful in many cases -- it would be good to have an option for `flow_from_directory` that allows one to choose whether all batches will have identical sizes or not.
